zkevm-circuits/src/keccak_circuit/keccak_packed.rs

```diff
@@ -0,0 +1,1770 @@
+        } else {
+            unreachable!();
+        }
+    }
+    parts
+}
+
+impl<F: Field> KeccakPackedConfig<F> {
+    pub(crate) fn configure(meta: &mut ConstraintSystem<F>, r: F) -> Self {
+        let q_enable = meta.selector();
+        let q_first = meta.selector();
+        let q_round = meta.fixed_column();
+        let q_absorb = meta.fixed_column();
+        let q_round_end = meta.selector();
+        let is_final = meta.advice_column();
+        let hash_rlc = meta.advice_column();
+        let state = array_init::array_init(|_| meta.advice_column());
+        let absorb_from = meta.advice_column();
+        let absorb_data = meta.advice_column();
+        let absorb_result = meta.advice_column();
+        let squeeze_packed = meta.advice_column();
+        let round_cst = meta.fixed_column();
+        let normalize_3 = array_init::array_init(|_| meta.lookup_table_column());
+        let normalize_4 = array_init::array_init(|_| meta.lookup_table_column());
+        let normalize_6 = array_init::array_init(|_| meta.lookup_table_column());
+        let chi_base_table = array_init::array_init(|_| meta.lookup_table_column());
+        let chi_ext_table = array_init::array_init(|_| meta.lookup_table_column());
+        let pack_unpack_table = array_init::array_init(|_| meta.lookup_table_column());
+
+        let mut cell_values = Vec::new();
+
+        let mut s = vec![vec![0u64.expr(); 5]; 5];
+        let mut s_next = vec![vec![0u64.expr(); 5]; 5];
+        let mut round_cst_expr = 0.expr();
+        meta.create_gate("Query state bits", |meta| {
+            let mut counter = 0;
+            for i in 0..5 {
+                for j in 0..5 {
+                    s[i][j] = meta.query_advice(state[counter], Rotation::cur());
+                    s_next[i][j] = meta.query_advice(state[counter], Rotation::next());
+                    counter += 1;
+                }
+            }
+            round_cst_expr = meta.query_fixed(round_cst, Rotation::cur());
+            vec![0u64.expr()]
+        });
+        let mut absorb_from_expr = 0u64.expr();
+        let mut absorb_data_expr = 0u64.expr();
+        let mut absorb_result_expr = 0u64.expr();
+
+        let mut absorb_from_next_expr = vec![0u64.expr(); NUM_WORDS_TO_ABSORB];
+        let mut absorb_data_next_expr = vec![0u64.expr(); NUM_WORDS_TO_ABSORB];
+        let mut absorb_result_next_expr = vec![0u64.expr(); NUM_WORDS_TO_ABSORB];
+        meta.create_gate("Query absorb data", |meta| {
+            absorb_from_expr = meta.query_advice(absorb_from, Rotation::cur());
+            absorb_data_expr = meta.query_advice(absorb_data, Rotation::cur());
+            absorb_result_expr = meta.query_advice(absorb_result, Rotation::cur());
+
+            for i in 0..NUM_WORDS_TO_ABSORB {
+                absorb_from_next_expr[i] = meta.query_advice(absorb_from, Rotation((i + 1) as i32));
+                absorb_data_next_expr[i] = meta.query_advice(absorb_data, Rotation((i + 1) as i32));
```

Comment:
I don't think many/large rotations have a big impact on the prover performance. The extended domain is currently x4 bigger, with https://github.com/privacy-scaling-explorations/halo2/pull/76 this will go down to just x2. This will give a 15% performance increase because FFTs will be smaller. So memory should also be improved quite a bit, but I haven't measured yet by how much.

I think it may also be possible to use copy constraints here instead, but I don't believe that will matter much.

Reply:
I like it! Thanks for the explanation!!